Refactor Academy router

RichDom2185 opened the pull request:
### Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

* Create `GuardedRoute` class for builder pattern
* Reformat app router config for readability
* Migrate Academy routes to use new Data Loader API
* Use lazy loading + code-split Academy routes for performance #2499
* Refactor routing logic (assessment configurations, not found behavior)

Please refer to commit history for details.

See #2507
See #2484

**TODO:**

- [x] Fix infinite render update loop when visiting admin panel
- [x] Fix tests (tried following https://reactrouter.com/en/main/routers/create-memory-router without luck)
- [x] Potentially refactor `routerConfig.tsx` to use `GuardedRoute` where possible

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Code quality improvements

### How to test

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. -->

Everything should still work.

### Checklist

<!-- Please delete options that are not relevant. -->

- [x] I have tested this code
- [ ] I have updated the documentation

chownces commented on src/pages/academy/academyRoutes.tsx:
```diff
@@ -0,0 +1,108 @@
+  { path: `teamformation/edit/${teamRegExp}`, lazy: TeamFormationForm },
+  { path: 'teamformation/import', lazy: TeamFormationImport },
+  { path: 'dashboard', lazy: Dashboard }
+].map(r => new GuardedRoute(r).check(s => s.session.role !== Role.Student, notFoundPath).build());
```
Let's change this to a whitelist instead.

RichDom2185 replied:
Resolved in 90dba2e1f75fad50fc5842915413181fa44dc5b0.

RichDom2185 commented:
> * [x]  Fix infinite render update loop when visiting admin panel
> * [x]  Fix tests (tried following https://reactrouter.com/en/main/routers/create-memory-router without luck)
> * [x]  Potentially refactor `routerConfig.tsx` to use `GuardedRoute` where possible

@chownces I've fixed the infinite rerender issue, and also refactored `ensureUserAndRole` to use `GuardedRoute`, thereby eliminating the dependency on `role`. I think in the future we should move towards a static router (thereby removing the need for `UPDATE_REACT_ROUTER` action) to decrease complexity – just access the store directly via `store.getState()` within loader functions whenever we need to have dynamic routing logic.

This will reduce the complexity of the implementation.

chownces reviewed (commented):
There's currently a bug in the academy routes when the route is non-existent. Instead of 404-ing, the page hangs (likely in an infinite loop).

Try something like `http://localhost:8000/courses/11/sourcecastt`

RichDom2185 commented:
> There's currently a bug in the academy routes when the route is non-existent. Instead of 404-ing, the page hangs (likely in an infinite loop).
> 
> Try something like `http://localhost:8000/courses/11/sourcecastt`

Somehow the new 404 behaviour seems to be causing the infinite loop. I've restored the original behaviour of rendering `<NotFound />` as a catch-all, instead of redirecting first to a `/not_found` path, then rendering the component there. This fixes the issue.

chownces reviewed (approved):
LGTM. Thanks
